Dockerfile

```diff
@@ -34,17 +34,24 @@ ENV CGO_ENABLED=0 \
     GOARM=${TARGETVARIANT} \
     LDFLAGS="-X ${PKG}/pkg/buildinfo.Version=${VERSION} -X ${PKG}/pkg/buildinfo.GitSHA=${GIT_SHA} -X ${PKG}/pkg/buildinfo.GitTreeState=${GIT_TREE_STATE} -X ${PKG}/pkg/buildinfo.ImageRegistry=${REGISTRY}"
 
+RUN mkdir -p /output/usr/bin
 WORKDIR /go/src/github.com/vmware-tanzu/velero
-
+COPY go.mod go.sum /go/src/github.com/vmware-tanzu/velero/
+# --mount=type=cache,target=/go/pkg/mod,id=vbb allows reuse of build cache across builds instead of invalidating whole cache when go.mod changes
+# id is to allow other stages to use the same cache path without conflicting with this stage.
+# velero-builder-helper and velero-builder share the same cache path and id to share the downloaded dependencies.
+# restic-builder uses a different cache path and id to avoid sharing cache with velero-builder-helper and velero-builder.
+RUN --mount=type=cache,target=/go/pkg/mod,id=vbb go mod download
 COPY . /go/src/github.com/vmware-tanzu/velero
 
-RUN mkdir -p /output/usr/bin && \
-    export GOARM=$( echo "${GOARM}" | cut -c2-) && \
-    go build -o /output/${BIN} \
-    -ldflags "${LDFLAGS}" ${PKG}/cmd/${BIN} && \
-    go build -o /output/velero-helper \
-    -ldflags "${LDFLAGS}" ${PKG}/cmd/velero-helper && \
-    go clean -modcache -cache
+
+FROM velero-builder-base AS velero-builder-helper
```

Comment:
Do we need to make this helper builder into a separate section?
By running it in the same section with velero, it will also reuse the same cache with Velero.
And we are also considering to remove this velero-helper binary and use velero with different subcommands instead.

Reply:
Right. This separate helper section is not caching related but perhaps help with using more CPU threads I thought. I can move it back together. Most of the time savings is eliminating download on minor code change.

Reply:
Reverted the separate sections. Please take a look. Thanks.